File: #0979.distribute-coins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int distributeCoins(TreeNode* root) {
        
        this->tree(root);
        return result;
    }

private:
    int result = 0;
    
    int tree(TreeNode* node) {

        if (node == NULL) {
            return 0;
        }
        
        int left = tree(node->left);
        int right = tree(node->right);
        result += abs(left) + abs(right);
        return node->val - 1 + left + right;
    }
};

File: #0979.distribute-coins-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var distributeCoins = function(root) {
    let result = 0;
    const move = (node) => {
        if (!node) {
            return 0;
        }
        
        const left = move(node.left);
        const right = move(node.right);
        result += Math.abs(left) + Math.abs(right);
        return left + right + node.val - 1;
    }
    
    move(root);
    return result;
};
